### PR TITLE
fix(try): try an anonymous request before throwing an auth error

### PR DESCRIPTION
**Pull request by jgravois**
AFFECTS PACKAGES:
@esri/arcgis-rest-auth
@esri/arcgis-rest-request

ISSUES CLOSED: #559

### CODE REVIEW

File: packages/arcgis-rest-request/src/request.ts [hunk 1]
```diff
@@ -139,11 +142,20 @@ export function request(
...
         err.url = url;
         err.options = options;
-        throw err;
+        /**
+         * if an attempt is made to talk to an unfederated server
+         * first try the request anonymously. if a 'token required'
+         * error is thrown, throw the UNFEDERATED error then.
+         */
+        originalAuthError = err;
+        return Promise.resolve("");
```
Comment (jgravois) [inline packages/arcgis-rest-request/src/request.ts]: initially i thought we'd need something like this:
```ts
if (err.code === "NOT_FEDERATED") {
  originalAuthError = err;
  return Promise.resolve("");
}
throw err
```
but in updating the tests i couldn't isolate a scenario in which `getToken()` throws a different error.

File: packages/arcgis-rest-request/src/utils/check-for-errors.ts [hunk 2]
```diff
@@ -35,7 +36,11 @@ export function checkForErrors(
       messageCode === "GWM_0003" ||
       (code === 400 && message === "Unable to generate token.")
     ) {
-      throw new ArcGISAuthError(message, errorCode, response, url, options);
+      if (originalAuthError) {
+        throw originalAuthError;
```
Comment (tomwayson) [inline packages/arcgis-rest-request/src/utils/check-for-errors.ts]: ah, great idea. this was my concern about re-trying, that we'd show the wrong error message.

**Comment by jgravois** [comment]
to do:

- [ ] cache the url as a 'trusted' server so that repeat requests won't trigger a federation check.

**Comment by jgravois** [comment]
in c0dad417c6f034ff4583958eba643427612ba8e9 i implemented logic to cache the root server url in the event of a successful anonymous request in order to avoid repeat federation checks.

ref: https://github.com/Esri/arcgis-rest-js/pull/423#issuecomment-449473077

**Comment by patrickarlt** [comment]
@jgravois new changes LGTM.

**Comment by jgravois** [comment]
thanks @patrickarlt!

after i merge this i'm going to cut a `v2.0.1` release.

**Pull request merged**